src/main/scala/tilelink/Broadcast.scala

```diff
@@ -159,119 +184,214 @@ class TLBroadcast(lineBytes: Int, numTrackers: Int = 4, bufferless: Boolean = fa
+      val matchTrackers = VecInit(trackers.map { t => t.line === in.a.bits.address >> lineShift }).asUInt
       val matchTracker = matchTrackers.orR()
       val allocTracker = freeTrackers & ~(leftOR(freeTrackers) << 1)
       val selectTracker = Mux(matchTracker, matchTrackers, allocTracker)
+      val trackerReadys = VecInit(trackers.map(_.in_a.ready)).asUInt
+      val trackerReady = (selectTracker & trackerReadys).orR()
 
-      val trackerReady = Vec(trackers.map(_.in_a.ready)).asUInt
-      in.a.ready := (!a_first || !probe_busy) && (selectTracker & trackerReady).orR()
+      in.a.ready := (!a_first || filter.io.request.ready) && trackerReady
       (trackers zip selectTracker.asBools) foreach { case (t, select) =>
-        t.in_a.valid := in.a.valid && select && (!a_first || !probe_busy)
+        t.in_a.valid := in.a.valid && select && (!a_first || filter.io.request.ready)
         t.in_a.bits := in.a.bits
         t.in_a_first := a_first
-        t.probe := (if (caches.size == 0) UInt(0) else Mux(a_cache.orR(), UInt(caches.size-1), UInt(caches.size)))
       }
 
-      val acq_perms = MuxLookup(in.a.bits.param, Wire(UInt(width = 2)), Array(
-        TLPermissions.NtoB -> TLPermissions.toB,
-        TLPermissions.NtoT -> TLPermissions.toN,
-        TLPermissions.BtoT -> TLPermissions.toN))
-
-      when (in.a.fire() && a_first) {
-        probe_todo  := ~a_cache // probe all but the cache who poked us
-        probe_line  := in.a.bits.address >> lineShift
-        probe_perms := MuxLookup(in.a.bits.opcode, Wire(UInt(width = 2)), Array(
-          TLMessages.PutFullData    -> TLPermissions.toN,
-          TLMessages.PutPartialData -> TLPermissions.toN,
-          TLMessages.ArithmeticData -> TLPermissions.toN,
-          TLMessages.LogicalData    -> TLPermissions.toN,
-          TLMessages.Get            -> TLPermissions.toB,
-          TLMessages.Hint           -> MuxLookup(in.a.bits.param, Wire(UInt(width = 2)), Array(
-            TLHints.PREFETCH_READ   -> TLPermissions.toB,
-            TLHints.PREFETCH_WRITE  -> TLPermissions.toN)),
-          TLMessages.AcquireBlock   -> acq_perms,
-          TLMessages.AcquirePerm    -> acq_perms))
+      val acq_needT = MuxLookup(in.a.bits.param, Wire(Bool()), Array(
+        TLPermissions.NtoB -> false.B,
+        TLPermissions.NtoT -> true.B,
+        TLPermissions.BtoT -> true.B))
+      val req_needT = MuxLookup(in.a.bits.opcode, Wire(Bool()), Array(
+        TLMessages.PutFullData    -> true.B,
+        TLMessages.PutPartialData -> true.B,
+        TLMessages.ArithmeticData -> true.B,
+        TLMessages.LogicalData    -> true.B,
+        TLMessages.Get            -> false.B,
+        TLMessages.Hint           -> MuxLookup(in.a.bits.param, Wire(Bool()), Array(
+          TLHints.PREFETCH_READ   -> false.B,
+          TLHints.PREFETCH_WRITE  -> true.B)),
+        TLMessages.AcquireBlock   -> acq_needT,
+        TLMessages.AcquirePerm    -> acq_needT))
```

Comment:
These seem like helper functions that could be defined on the `edge`. They're globally true for all TL1.9 caching adapters that have to deal with these opcode/param types, right?

Reply:
Sure. Fixed.